tests/test_controller.py

```diff
@@ -387,29 +427,196 @@ def test_remove_items(self):
+        del no_id_doc['id']
+        no_id_response = MockRequestsResponse(401, content=json.dumps(no_id_doc))
+        self.http.responses.append(no_id_response)
+
+        request_args = self.create_register_request_args(document_url)
+        with self.app.test_request_context('/', **request_args):
+            response = self.controller.register(do_get=self.http.do_get)
+        assert_invalid_auth_document(response, 'is missing an id')
+
+        # A ProblemDetail is returned when the authentication document id
+        # doesn't match the submitted OPDS url.
+        mock_doc_response = MockRequestsResponse(401, content=mock_auth_doc)
+        self.http.responses.append(mock_doc_response)
+
+        url = 'https://fake.opds/okay/authentication_document'
+        request_args = self.create_register_request_args(url)
+        with self.app.test_request_context('/', **request_args):
+            response = self.controller.register(do_get=self.http.do_get)
+        assert_invalid_auth_document(response, "doesn't match submitted url")
+
+        # A ProblemDetail is returned when the authentication document doesn't
+        # have an RSA public key.
+        no_key_json = json.loads(mock_auth_doc)
+        del no_key_json['public_key']
+        no_public_key_response = MockRequestsResponse(
+            401, content=json.dumps(no_key_json)
+        )
+        self.http.responses.append(no_public_key_response)
+
+        request_args = self.create_register_request_args(document_url)
+        with self.app.test_request_context('/', **request_args):
+            response = self.controller.register(do_get=self.http.do_get)
+        assert_invalid_auth_document(response, "missing an RSA public_key")
+
+        # There's a key, but the type isn't RSA.
+        no_key_json['public_key'] = dict(type='safe', value='value')
+        no_public_key_response.content = json.dumps(no_key_json)
+        self.http.responses.append(no_public_key_response)
+
+        request_args = self.create_register_request_args(document_url)
+        with self.app.test_request_context('/', **request_args):
+            response = self.controller.register(do_get=self.http.do_get)
+        assert_invalid_auth_document(response, "missing an RSA public_key")
+
+        # There's an RSA public_key property, but there's no value there.
+        no_key_json['public_key']['type'] = 'RSA'
+        del no_key_json['public_key']['value']
+        no_public_key_response.content = json.dumps(no_key_json)
+        self.http.responses.append(no_public_key_response)
+
+        request_args = self.create_register_request_args(document_url)
+        with self.app.test_request_context('/', **request_args):
+            response = self.controller.register(do_get=self.http.do_get)
+        assert_invalid_auth_document(response, "missing an RSA public_key")
+
+    def test_register_succeeds_with_valid_authentication_document(self):
```

Comment:
If I'm reading this test right, you're using the `public_key` of an authentication document (associated with a particular library) to sign the shared secret for the entire circulation manager. At first I thought this was totally wrong, but upon reflection I see how this works. It doesn't matter which key you use to sign the shared secret with, so long as it's one the circulation manager has access to, and this lets you get rid of the [metadata key document](https://github.com/NYPL-Simplified/Simplified/wiki/Metadata-Wrangler#metadata-key-document) entirely. However, it is a little confusing, because you're mixing library-level stuff with server-level stuff.